### PR TITLE
Added ability to convert from aws-c-http http request object to java …

### PR DESCRIPTION
**Pull request by JonathanHenson**
…http request, and httprequest is now mutable in java land.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/software/amazon/awssdk/crt/http/HttpRequest.java [hunk 1]
```diff
@@ -56,10 +61,53 @@ public String getEncodedPath() {
...
 
+    public void addHeader(final HttpHeader header) {
+        headers.add(header);
+    }
+
+    public void addHeader(final String headerName, final String headerValue) {
+        headers.add(new HttpHeader(headerName, headerValue));
+    }
+
+    public void addHeaders(final HttpHeader headers) {
```
Comment (ColdenCullen) [inline src/main/java/software/amazon/awssdk/crt/http/HttpRequest.java]: Should this be an array or something?

File: src/main/java/software/amazon/awssdk/crt/auth/signing/AwsSigner.java [hunk 2]
```diff
@@ -25,7 +26,10 @@
         CompletableFuture<HttpRequest> future = new CompletableFuture<HttpRequest>();
 
         try {
-            awsSignerSignRequest(request, config, future);
+            //JNI is much easier to deal with primitives, so go ahead and hoist the headers into an array.
+            HttpHeader[] tempHeadersToSign = new HttpHeader[request.getHeaders().size()];
+            tempHeadersToSign = request.getHeaders().toArray(tempHeadersToSign);
+            awsSignerSignRequest(request, tempHeadersToSign, config, future);
```
Comment (graebm) [inline src/main/java/software/amazon/awssdk/crt/auth/signing/AwsSigner.java]: trivial: worth adding a HttpRequest.getHeadersArray() or a helper function somewhere?

File: src/main/java/software/amazon/awssdk/crt/http/HttpRequest.java [hunk 3]
```diff
@@ -56,10 +61,52 @@ public String getEncodedPath() {
...
+        this.headers.add(headers);
+    }
+
+    public void addQueryParam(final String paramName, final String paramValue, boolean urlEncode) {
```
Comment (graebm) [inline src/main/java/software/amazon/awssdk/crt/http/HttpRequest.java]: nuke this function if we don't need it now. no other language has this

File: src/main/java/software/amazon/awssdk/crt/http/HttpRequest.java (hunk 1 above)
Comment (JonathanHenson) [reply src/main/java/software/amazon/awssdk/crt/http/HttpRequest.java]: fixed.

File: src/native/http_request_utils.h [hunk 4]
```diff
@@ -36,6 +36,8 @@ struct aws_http_message *aws_http_request_new_from_java_http_request(
     jobjectArray jni_headers,
     jobject jni_body_stream);
 
+jobject aws_java_http_request_from_native(JNIEnv *env, struct aws_http_message *message);
```
Comment (graebm) [inline src/native/http_request_utils.h]: if this fails, a java exception has been set, right? Add a comment to that effect
Looking in the function, it does seem that the only things that can fail are the things that create java objects

File: src/native/http_request_utils.h [hunk 5]
```diff
@@ -36,6 +36,8 @@ struct aws_http_message *aws_http_request_new_from_java_http_request(
...
+
```
Comment (graebm) [inline src/native/http_request_utils.h]: We'll also need a function that modifies an existing native aws_http_request to match the modifications made to the Java HtttpRequest? Can you add that in this PR please?

File: src/main/java/software/amazon/awssdk/crt/auth/signing/AwsSigner.java (hunk 2 above)
Comment (JonathanHenson) [reply src/main/java/software/amazon/awssdk/crt/auth/signing/AwsSigner.java]: i can do that

File: src/main/java/software/amazon/awssdk/crt/http/HttpRequest.java (hunk 3 above)
Comment (JonathanHenson) [reply src/main/java/software/amazon/awssdk/crt/http/HttpRequest.java]: removed

**Pull request merged**